src/gfx/helpers/translation_gizmo.hpp

```diff
@@ -0,0 +1,113 @@
+
+#include "gizmos.hpp"
+
+namespace gfx {
+namespace gizmos {
+struct TranslationGizmo : public IGizmo, public IGizmoCallbacks {
+  float4x4 transform = linalg::identity;
+
+  Handle handles[3];
+  float4x4 dragStartTransform;
+  float3 dragStartPoint;
+
+  float scale = 1.0f;
+  const float axisRadius = 0.1f;
+  const float axisLength = 0.7f;
+
+  float getGlobalAxisRadius() const { return axisRadius * scale; }
+  float getGlobalAxisLength() const { return axisLength * scale; }
+
+  TranslationGizmo() {
+    for (size_t i = 0; i < 3; i++) {
+      handles[i].userData = (void *)i;
+      handles[i].callbacks = this;
+    }
+  }
+
+  void update(InputContext &inputContext) {
+    for (size_t i = 0; i < 3; i++) {
+      auto &handle = handles[i];
+
+      float3 fwd{};
+      fwd[i] = 1.0f;
+      float3 t1 = float3(-fwd.z, -fwd.x, fwd.y);
+      float3 t2 = linalg::cross(fwd, t1);
+
+      auto &min = handle.selectionBox.min;
+      auto &max = handle.selectionBox.max;
+      min = -t1 * getGlobalAxisRadius() - t2 * getGlobalAxisRadius();
+      max = t1 * getGlobalAxisRadius() + t2 * getGlobalAxisRadius() + fwd * getGlobalAxisLength();
+
+      handle.selectionBoxTransform = transform;
+
+      inputContext.updateHandle(handle);
+    }
+  }
+
+  size_t getHandleIndex(Handle &inHandle) { return size_t(inHandle.userData); }
+
+  float3 getAxisDirection(size_t index, float4x4 transform) {
+    float3 base{};
+    base[index] = 1.0f;
+    return linalg::normalize(linalg::mul(transform, float4(base, 0)).xyz());
+  }
+
+  virtual void grabbed(InputContext &context, Handle &handle) {
+    dragStartTransform = transform;
+
+    size_t index = getHandleIndex(handle);
+    SPDLOG_DEBUG("Handle {} ({}) grabbed", index, getAxisDirection(index, dragStartTransform));
+
+    dragStartPoint = hitOnPlane(context.eyeLocation, context.rayDirection, extractTranslation(dragStartTransform),
+                                getAxisDirection(index, dragStartTransform));
+  }
+
+  virtual void released(InputContext &context, Handle &handle) {
+    size_t index = getHandleIndex(handle);
+    SPDLOG_DEBUG("Handle {} ({}) released", index, getAxisDirection(index, dragStartTransform));
+  }
+
+  virtual void move(InputContext &context, Handle &inHandle) {
+    float3 fwd = getAxisDirection(getHandleIndex(inHandle), dragStartTransform);
+    float3 hitPoint = hitOnPlane(context.eyeLocation, context.rayDirection, dragStartPoint, fwd);
+
+    float3 delta = hitPoint - dragStartPoint;
+    delta = linalg::dot(delta, fwd) * fwd;
+    transform = linalg::mul(linalg::translation_matrix(delta), dragStartTransform);
+  }
+
+  void render(InputContext &inputContext, GizmoRenderer &renderer) {
+    for (size_t i = 0; i < 3; i++) {
+      auto &handle = handles[i];
+
+      // Debug draw
+      float4 color = float4(.7, .7, .7, 1.);
+      uint32_t thickness = 1;
+      bool hovered = inputContext.hovered && inputContext.hovered == &handle;
```

Comment:
Nit: hovering? Unless the action has occurred and concluded. If the user hovers above the control for 2 seconds and moves away is the debug draw still on?